File: invoice-agent-system-java/src/main/java/com/example/invoice/integrations/OpenAiHttpClient.java
package com.example.invoice.integrations;

import com.example.invoice.config.InvoiceProperties;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.springframework.stereotype.Component;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.time.Duration;
import java.util.List;
import java.util.Map;

@Component
public class OpenAiHttpClient {
    private final InvoiceProperties properties;
    private final HttpClient httpClient;
    private final ObjectMapper mapper;

    public OpenAiHttpClient(InvoiceProperties properties, HttpClient httpClient, ObjectMapper mapper) {
        this.properties = properties;
        this.httpClient = httpClient;
        this.mapper = mapper;
    }

    public String chat(String systemPrompt, String userPrompt) throws Exception {
        InvoiceProperties.OpenAi cfg = properties.ai().openai();
        if (cfg.apiKey().isBlank()) {
            return "{}";
        }

        Map<String, Object> body = Map.of(
                "model", cfg.model(),
                "temperature", 0.0,
                "response_format", Map.of("type", "json_object"),
                "messages", List.of(
                        Map.of("role", "system", "content", systemPrompt),
                        Map.of("role", "user", "content", userPrompt)
                )
        );

        String url = trimTrailingSlash(cfg.baseUrl()) + "/v1/chat/completions";
        HttpRequest request = HttpRequest.newBuilder(URI.create(url))
                .timeout(Duration.ofSeconds(cfg.timeoutSeconds()))
                .header("Authorization", "Bearer " + cfg.apiKey())
                .header("Content-Type", "application/json")
                .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body)))
                .build();

        HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
        if (response.statusCode() < 200 || response.statusCode() > 299) {
            throw new IllegalStateException("OpenAI-compatible endpoint failed: HTTP " + response.statusCode() + " " + abbreviate(response.body(), 500));
        }
        JsonNode root = mapper.readTree(response.body());
        JsonNode choices = root.path("choices");
        return choices.isArray() && !choices.isEmpty()
                ? choices.get(0).path("message").path("content").asText("{}")
                : "{}";
    }

    private static String trimTrailingSlash(String value) {
        return value.endsWith("/") ? value.substring(0, value.length() - 1) : value;
    }

    private static String abbreviate(String value, int max) {
        if (value == null || value.length() <= max) {
            return value;
        }
        return value.substring(0, max);
    }
}

File: invoice-agent-system/src/main/kotlin/com/example/invoice/integrations/OpenAiHttpClient.kt
package com.example.invoice.integrations

import com.example.invoice.config.InvoiceProperties
import com.fasterxml.jackson.databind.ObjectMapper
import org.springframework.stereotype.Component
import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

@Component
class OpenAiHttpClient(
    private val properties: InvoiceProperties,
    private val httpClient: HttpClient,
    private val mapper: ObjectMapper
) {
    fun chat(systemPrompt: String, userPrompt: String): String {
        val cfg = properties.ai.openai
        if (cfg.apiKey.isBlank()) {
            return "{}"
        }
        val body = mapOf(
            "model" to cfg.model,
            "temperature" to 0.0,
            "response_format" to mapOf("type" to "json_object"),
            "messages" to listOf(
                mapOf("role" to "system", "content" to systemPrompt),
                mapOf("role" to "user", "content" to userPrompt)
            )
        )
        val url = cfg.baseUrl.trimEnd('/') + "/v1/chat/completions"
        val request = HttpRequest.newBuilder(URI.create(url))
            .timeout(Duration.ofSeconds(cfg.timeoutSeconds))
            .header("Authorization", "Bearer ${cfg.apiKey}")
            .header("Content-Type", "application/json")
            .POST(HttpRequest.BodyPublishers.ofString(mapper.writeValueAsString(body)))
            .build()
        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        if (response.statusCode() !in 200..299) {
            throw IllegalStateException("OpenAI-compatible endpoint failed: HTTP ${response.statusCode()} ${response.body().take(500)}")
        }
        val root = mapper.readTree(response.body())
        return root.path("choices").firstOrNull()?.path("message")?.path("content")?.asText() ?: "{}"
    }
}
